LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs

```diff
@@ -67,32 +65,30 @@ private string Message(DownlinkPktFwdMessage message, Region region)
             writer.WriteNumber("dC", (int)deviceClassType);
 
             // Getting and writing payload bytes
-            var pduBytes = Convert.FromBase64String(message.Txpk.Data);
+            var pduBytes = message.Data;
             var pduChars = new char[pduBytes.Length * 2];
             Hexadecimal.Write(pduBytes, pduChars);
             writer.WriteString("pdu", pduChars);
 
 #pragma warning disable CA5394 // Do not use insecure randomness. This is fine as not used for any crypto operations.
-            writer.WriteNumber("diid", this.random.Next(int.MinValue, int.MaxValue));
+            var diid = this.random.Next(int.MinValue, int.MaxValue);
+            writer.WriteNumber("diid", diid);
 #pragma warning restore CA5394 // Do not use insecure randomness
-
-#pragma warning disable CS0618 // Type or member is obsolete
-            var dataRate = region.GetDRFromFreqAndChan(message.Txpk.Datr);
-#pragma warning restore CS0618 // Type or member is obsolete
+            LogSendingMessage(this.logger, message.StationEui, diid, ConversionHelper.ByteArrayToString(message.Data), null);
 
             if (deviceClassType is LoRaDeviceClassType.A)
             {
                 writer.WriteNumber("RxDelay", message.LnsRxDelay);
-                writer.WriteNumber("RX1DR", (int)dataRate);
-                writer.WriteNumber("RX1Freq", (ulong)(message.Txpk.Freq * 1e6));
-                writer.WriteNumber("RX2DR", (int)region.GetDefaultRX2ReceiveWindow().DataRate);
-                writer.WriteNumber("RX2Freq", (ulong)region.GetDefaultRX2ReceiveWindow().Frequency);
+                writer.WriteNumber("RX1DR", (int)message.DataRateRx1);
+                writer.WriteNumber("RX1Freq", message.FrequencyRx1.AsUInt64);
+                writer.WriteNumber("RX2DR", (int)message.DataRateRx2);
+                writer.WriteNumber("RX2Freq", message.FrequencyRx2.AsUInt64);
                 writer.WriteNumber("xtime", message.Xtime);
             }
             else if (deviceClassType is LoRaDeviceClassType.C)
             {
-                writer.WriteNumber("RX2DR", (int)dataRate);
-                writer.WriteNumber("RX2Freq", (ulong)(message.Txpk.Freq * 1e6));
+                writer.WriteNumber("RX2DR", (int)message.DataRateRx2);
+                writer.WriteNumber("RX2Freq", message.FrequencyRx2.AsUInt64);
```

Comment:
Case as in the DR case:

```suggestion
                writer.WriteNumber("RX2Freq", (ulong)message.FrequencyRx2);
```